fix: 공시 상세 조회 예외처리

bakgom123 opened the pull request:
### Description

<!--
  간단하게 PR의 목적을 설명하세요.
  이 PR이 해결하려는 문제나 추가하려는 기능에 대해 요약해주세요.
-->

## 공시 상세 조회

# 1. 기본 정상 케이스

- [x]  /v1/gongsi/123

# 2. 유효하지 않은 gongsiId 값

- [x]  /v1/gongsi/-1 (401 엑세스 토큰을 넣어주세요 에러가 뜸)
- [x]  /v1/gongsi/0
- [x]  /v1/gongsi/99999999999
- [x]  /v1/gongsi/abc (401 엑세스 토큰을 넣어주세요 에러가 뜸)

# 3. 경로 변수 누락

- [x]  /v1/gongsi/ (401 엑세스 토큰을 넣어주세요 에러가 뜸)

# 4. 특수 문자 포함

- [x]  /v1/gongsi/123%20 (401 엑세스 토큰을 넣어주세요 에러가 뜸)
- [x]  /v1/gongsi/123; (401 엑세스 토큰을 넣어주세요 에러가 뜸)

# 6. 여러 세그먼트 시도

- [x]  /v1/gongsi/123/456 (401 엑세스 토큰을 넣어주세요 에러가 뜸)

# 7. 다른 HTTP 메소드 시도 (물론 GET만 허용된다면)

- [x]  # PUT, POST, DELETE 등도 시도 (401 엑세스 토큰을 넣어주세요 에러가 뜸)

필요한 예외처리 끝났습니다.
JWTAuthFilter에서 gongsi쪽 검증을 통과하게 만들어서 401이 일어나는 이슈를 해결했습니다.
또한 gongsiId가 string인 예외적인 부분에 대한 추가 검증도 끝났습니다.

### Related Issues

<!--
  관련된 이슈 번호를 참고하세요.
  예: Fixes #123, Closes #456
-->
- Resolves #164 

### Changes Made

<!--
  이 PR에서 변경된 사항을 설명하세요.
  코드, 문서, 설정 등 변경된 내용을 상세히 기술합니다.
-->


### Screenshots or Video

<!--
  변경된 사항이 UI에 영향을 미치는 경우, 변경 전후의 스크린샷이나 동영상을 첨부하세요.
-->

### Testing

<!--
  변경 사항을 테스트한 방법을 설명하세요.
  테스트한 환경 (OS, 브라우저, 장치 등)과 테스트 절차를 구체적으로 기술합니다.
-->

POSTMAN을 통한 테스트 끝났습니다.

### Checklist

<!--
  PR 작성 시 다음 항목들을 확인하세요.
-->

- [ ] 코드가 정상적으로 컴파일되는지 확인했습니다.
- [ ] 모든 테스트가 성공적으로 통과했습니다.
- [ ] 관련 문서를 업데이트했습니다.
- [ ] PR이 관련 이슈를 정확히 참조하고 있습니다.
- [ ] 코드 스타일 가이드라인을 준수했습니다.
- [ ] 코드 리뷰어를 지정했습니다.

### Additional Notes

<!--
  리뷰어가 이해하는 데 도움이 될 추가적인 참고 사항이나 정보가 있다면 여기에 작성하세요.
-->

bakgom123 committed:
fix: 공시 상세 조회 예외처리

Unagi-zoso commented on src/main/resources/application-local.yml:
```diff
@@ -39,4 +45,4 @@ korean:
 jwt:
   secret-key: dflknlsdfjlskfncdsjlkmxnfjsklcxmvnfddddddddc
-  expiration-time: 604800000
+  expiration-time: 604800000
```
파일 막줄에 개행 넣으면 좋을거 같아요! Intellij 에선 관련 옵션도 있답니다.

Unagi-zoso commented on src/main/java/org/bob/siungongsi/controller/GongsiController.java:
```diff
@@ -58,7 +58,7 @@ public ResponseEntity<ApiResponseWrapper<GongsiListResponse>> getGongsiList(
   public ResponseEntity<ApiResponseWrapper<GongsiDetailResponse>> getGongsiDetail(
       @Parameter(description = "공시 ID", example = "101", required = true) @PathVariable("gongsiId")
-          Long gongsiId,
+          String gongsiId,
```
혹시 정수외 값처리가 힘들어 이렇게 한것일까요?!

Spring validation 과 전역 예외 핸들러를 사용한다면 비즈니스 코드에선 Long 을 유지할 수 있게 할 수 있답니다. 그 편이 더 좋을거 같아요!

bakgom123 replied:
네 예를 들어 abc 같은 공시 id가 들어온다면 그 부분에서 예외처리가 어려웠습니다. 
말씀해 주신 Spring validation부분을 조금 더 알아보고 적용해 볼 예정입니다. 리뷰 감사합니다!

Unagi-zoso commented on src/main/java/org/bob/siungongsi/dto/ApiResponseCode.java:
```diff
@@ -71,6 +71,8 @@ public enum ApiResponseCode {
       HttpStatus.INTERNAL_SERVER_ERROR, 5408, "notification_creation_inconsistency"),
 
+  RESOURCE_NOT_FOUND(HttpStatus.NOT_FOUND, 8404, "cannot_find_resource"),
+
```
다른 not found 는 메시지를
xxx_not_found 로 하셨는데 여기선 cannot_find_resource 로 하신 이유가 있으실까요?

bakgom123 replied:
resource not found가 일관성이 있어 보이고 더 좋은 것 같네요 당시에 빠르게 테스트해보려고 추가한 부분이라 이 부분도 수정하겠습니다.

Unagi-zoso commented on src/main/java/org/bob/siungongsi/security/JwtAuthFilter.java:
```diff
@@ -31,7 +31,7 @@ protected void doFilterInternal(
     String authHeader = request.getHeader("Authorization");
 
-    if (request.getRequestURI().matches("/v1/gongsi/\\d+")) {
+    if (request.getRequestURI().startsWith("/v1/gongsi/")) {
```
혹시 이건 어떤 문제 때문에 변경된건가엽?!

bakgom123 replied:
정규식을 `/v1/gongsi/\\d+` 이렇게 작성을 하게 된다면 `여러 세그먼트`가 나오는 상황, `음수 id`가 들어가는 상황, `문자열 id`가 들어가는 상황에선 인증을 하라는 응답을 받게 됩니다. (401 error code)
`gongsi` 쪽은 인증이 전혀 필요가 없기에 저 방식으로 변경을 해서 처리를 했습니다.

Unagi-zoso commented on src/main/java/org/bob/siungongsi/service/GongsiService.java:
```diff
@@ -63,10 +63,27 @@ private Long getCurrentUserId() {
   }
 
-  public Long getCompanyIdByGongsiId(Long gongsiId) {
+  private Long parseGongsiId(String gongsiId) {
```
Controller 파라미터에서 Long 으로 바로 받는게 좋아보여욧

bakgom123 replied:
해당 부분도 문자열 id를 처리하는 데 있어서 어려움이 있었습니다. 조금 더 알아보고 Long으로 받지만 문자열에 대한 예외처리가 될 수 있는 방법을 찾아보겠습니다.

Unagi-zoso reviewed (commented):
센한테 말한거랑 같은거 같은데 controller 에서 String 으로 받지 말고 Long 으로 받는게 좋아보여요!
유효성 검사가 걱정이라면 Spring validation 과 전역 예외 핸들러를 사용하는게 좋아보여요.

bakgom123 committed:
fix: gongsiId String -> Long으로 변경

bakgom123 committed:
fix: conflict 해결

dslov89 commented on src/main/java/org/bob/siungongsi/exception/GlobalExceptionHandler.java:
```diff
@@ -40,32 +38,19 @@ public ApiResponseWrapper handleNullPointerException(NullPointerException ex) {
-    return ResponseEntity.status(ApiResponseCode.API_WRONG_TYPE_REQUEST.getHttpStatus())
-        .body(ApiResponseWrapper.error(ApiResponseCode.API_WRONG_TYPE_REQUEST));
-  }
-
```
이 부분 왜 삭제하셨나요?

bakgom123 committed:
fix: conflict 해결중 삭제된 코드 복구

bakgom123 committed:
Merge branch 'develop' into fix/gongsidetail

bakgom123 committed:
fix: test 실패해서 다시 커밋

bakgom123 committed:
fix: global exception 수정

bakgom123 committed:
fix: 글로벌 예외 처리 병합

bakgom123 committed:
fix: admincontroller 테스트 삭제

bakgom123 committed:
fix: ConstraintViolationException 하나로 통합

bakgom123 committed:
fix: 체계적으로 예외 처리를 수정

bakgom123 committed:
chore: 스프링 404 예외 처리 및 정적 리소스 설정 변경

Unagi-zoso committed:
fix: 잘못된 파라미터명 수정

api 에서 사용되는 정렬옵션 파라미터명은 sort 이다. (sortType -> sort)

Unagi-zoso committed:
feat: MissingPathVariableException 핸들러 작성

PathVariable 누락 시 예외 핸들러를 추가했습니다. fieldToCodeMap 을 응용해 구체적인 응답을 던져줄까 싶은데 아직 요구가 없어 일반적은 응답코드를 반환합니다. (API_BAD_REQUEST)

Unagi-zoso committed:
ci: 최신 브랜치 반영

Unagi-zoso commented on src/main/java/org/bob/siungongsi/exception/GlobalExceptionHandler.java:
```diff
@@ -61,15 +61,16 @@ public ApiResponseWrapper handleNullPointerException(NullPointerException ex) {
-          .body(ApiResponseWrapper.error(ApiResponseCode.GONGSI_INVALID_GONGSI_ID));
+    ApiResponseCode code = fieldToCodeMap.get(paramName);
+    if (code != null) {
+      return ResponseEntity.status(code.getHttpStatus()).body(ApiResponseWrapper.error(code));
```
커밋 분리를 잘못했네요. 이 작업은 fieldToCodeMap 을 이용해 전역 핸들러에 일반화적인 요소를 추가하려한 작업입니다.

Unagi-zoso reviewed (approved):
좋습니닷 👍
